### DISCUSSION TITLE
Laravel Migration events don't fire

### DISCUSSION THREAD

**Discussion by gasparyanm** [laravel/framework#32778] [General] [answered]
Laravel Migration event listeners don't work in my case.

**config/app.php**

```php
'providers' => [
...
 App\Providers\EventServiceProvider::class,
 ....
```

```php
namespace App\Providers;

use App\Listeners\DeleteUnitsAwsImages;
use Illuminate\Database\Events\MigrationsStarted;
use Illuminate\Foundation\Support\Providers\EventServiceProvider as ServiceProvider;
use Illuminate\Support\Facades\Event;

class EventServiceProvider extends ServiceProvider
{
    /**
     * The event listener mappings for the application.
     *
     * @var array
     */
    protected $listen = [
        MigrationsStarted::class => [
            DeleteUnitsAwsImages::class,
        ]
    ];
....
```

**Listener**
```php
<?php

namespace App\Listeners;

use Illuminate\Foundation\Events\Dispatchable;

class DeleteUnitsAwsImages
{
    use Dispatchable;

    public function __construct()
    {
        \Log::channel('daily')->info(['class' => '11111111111']);
    }

    public function handle()
    {
        \Log::channel('daily')->info(['class' => '22222222']);

    }
}
```

**Comment by ahinkle** [comment]
What events are registered? You can see this by running `php artisan event:list`

**Reply by gasparyanm** [reply]
```php
+----------------------------------------------+------------------------------------+
| Event                                        | Listeners                          |
+----------------------------------------------+------------------------------------+
| Illuminate\Database\Events\MigrationsStarted | App\Listeners\DeleteUnitsAwsImages |
+----------------------------------------------+------------------------------------+
```

This call as separate function call wworks
```\App\Listeners\DeleteUnitsAwsImages::dispatch();```
But I need to make working Listener after migration started.
Actually I need to catch migrate:fresh event and make listener to run additional script.
Need to open an issue ?

**Comment by lupinitylabs** [comment, marked as answer]
@M45678 Your code above works flawlessly for me on a fresh Laravel 7.11.0.
Have you made sure your logging setup works? Try putting a `dd('Test')` in the `handle()` method of `DeleteUnitsAwsImages` and do an `artisan migrate fresh` to test the setup.

**Reply by gasparyanm** [reply]
Yes, I've tried with dd, but nothing in Listener works. 
May it be from anything else?

**Reply by lupinitylabs** [reply]
One other thing I noticed is that you use `Illuminate\Foundation\Events\Dispatchable`, which is a trait for events, not for listeners. If you mean to dispatch it on the queue like a job, it should implement ShouldQueue and have the `Illuminate\Foundation\Bus\Dispatchable` trait?
However, it still works normally for me as long as I don't use dispatching.
Have you tried listening to the MigrationsStarted event on a fresh installation of Laravel? You opened an issue, so you obviously think there is a bug in the framework - but I cannot confirm that.
Which version of Laravel are you using on your project by the way?
It's hard to say more without knowing the complete code of the project...

**Reply by gasparyanm** [reply]
Thank you for your reply.
I've compared my project code and fresh Laravel code.
The reason was overridden **MigrationServiceProvider**  .
there missed **events** parameter when creating **Migrator** class.
Here the changed / working class

```php
namespace App\Providers;

use App\Migrator;
use Illuminate\Foundation\Application;
use Illuminate\Support\ServiceProvider;

class MigrationServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->singleton('migrator', function (Application $app) {
            return new Migrator(
                $app->make("migration.repository"),
                $app->make("db"),
                $app->make("files"),
                $app->make('events')// missed in old code
            );
        });
    }
} 
```
